File: whats-on-rva/scripts/import-grtc-gtfs.mjs
/**
 * Downloads GRTC static GTFS and writes trimmed JSON for the app.
 * Run: npm run import-grtc  (requires network; re-run when schedules change)
 *
 * Feed terms: https://www.ridegrtc.com/gtfs-files-developer-license-agreement-and-terms-of-use
 */
import AdmZip from 'adm-zip';
import { writeFileSync, mkdirSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, '..');
const DATA = join(ROOT, 'src', 'data');

const GTFS_URL = 'https://www.ridegrtc.com/wp-content/uploads/2025/02/gtfs.zip';

/** Match `src/lib/richmondBounds.js` city box */
const BOUNDS = { south: 37.465, north: 37.615, west: -77.585, east: -77.375 };

function inBounds(lat, lng) {
  return lat >= BOUNDS.south && lat <= BOUNDS.north && lng >= BOUNDS.west && lng <= BOUNDS.east;
}

/** Minimal CSV row parser (handles quoted fields). */
function parseCsvRow(line) {
  const out = [];
  let cur = '';
  let inQ = false;
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (c === '"') {
      inQ = !inQ;
    } else if (c === ',' && !inQ) {
      out.push(cur);
      cur = '';
    } else {
      cur += c;
    }
  }
  out.push(cur);
  return out.map((s) => s.trim());
}

function parseStops(text) {
  const lines = text.split(/\r?\n/).filter(Boolean);
  const header = lines[0].split(',');
  const idx = {
    id: header.indexOf('stop_id'),
    code: header.indexOf('stop_code'),
    name: header.indexOf('stop_name'),
    lat: header.indexOf('stop_lat'),
    lon: header.indexOf('stop_lon'),
  };
  const rows = [];
  for (let i = 1; i < lines.length; i++) {
    const cols = parseCsvRow(lines[i]);
    if (cols.length < 5) continue;
    const lat = parseFloat(cols[idx.lat]);
    const lng = parseFloat(cols[idx.lon]);
    if (!Number.isFinite(lat) || !Number.isFinite(lng)) continue;
    rows.push({
      id: String(cols[idx.id] ?? ''),
      code: cols[idx.code] ? String(cols[idx.code]) : '',
      name: cols[idx.name] || 'Stop',
      lat,
      lng,
    });
  }
  return rows;
}

async function main() {
  mkdirSync(DATA, { recursive: true });
  console.log('Fetching', GTFS_URL);
  const res = await fetch(GTFS_URL);
  if (!res.ok) throw new Error(`GTFS download failed: ${res.status}`);
  const zip = new AdmZip(Buffer.from(await res.arrayBuffer()));

  const read = (name) => {
    const e = zip.getEntry(name);
    if (!e) throw new Error(`Missing ${name} in zip`);
    return e.getData().toString('utf8');
  };

  const feedInfoTxt = read('feed_info.txt');
  const feedLines = feedInfoTxt.split(/\r?\n/).filter(Boolean);
  const feedCols = parseCsvRow(feedLines[1] || '');
  const feedVersion = feedCols[5] || 'unknown';
  const feedEnd = feedCols[4] || '';

  const stops = parseStops(read('stops.txt'));
  const inBox = stops.filter((s) => inBounds(s.lat, s.lng));

  const routesTxt = read('routes.txt');
  const routeLines = routesTxt.split(/\r?\n/).filter(Boolean);
  const rHeader = routeLines[0].split(',');
  const pulseRouteIds = new Set();
  for (let i = 1; i < routeLines.length; i++) {
    const c = parseCsvRow(routeLines[i]);
    const routeId = c[0];
    const shortName = c[2] || '';
    const longName = (c[3] || '').toUpperCase();
    if (routeId === 'BRT' || shortName === 'BRT' || longName.includes('PULSE')) {
      pulseRouteIds.add(routeId);
    }
  }

  const tripsTxt = read('trips.txt');
  const tripLines = tripsTxt.split(/\r?\n/).filter(Boolean);
  const brtTripIds = new Set();
  for (let i = 1; i < tripLines.length; i++) {
    const c = parseCsvRow(tripLines[i]);
    const tripId = c[0];
    const routeId = c[1];
    if (pulseRouteIds.has(routeId)) brtTripIds.add(tripId);
  }

  console.log('BRT/Pulse trips:', brtTripIds.size);

  const stopTimesTxt = read('stop_times.txt');
  const pulseStopIds = new Set();
  const stLines = stopTimesTxt.split(/\r?\n/);
  for (let i = 1; i < stLines.length; i++) {
    const line = stLines[i];
    if (!line.trim()) continue;
    const c = parseCsvRow(line);
    const tripId = c[0];
    if (!brtTripIds.has(tripId)) continue;
    const stopId = String(c[3] ?? '');
    if (stopId) pulseStopIds.add(stopId);
  }

  const stopById = new Map(stops.map((s) => [s.id, s]));
  const pulseStops = [];
  for (const sid of pulseStopIds) {
    const s = stopById.get(sid);
    if (s && inBounds(s.lat, s.lng)) {
      pulseStops.push({
        id: `pulse-${s.id}`,
        stopId: s.id,
        code: s.code,
        label: `Pulse (BRT) · ${s.name}`,
        lat: s.lat,
        lng: s.lng,
      });
    }
  }

  pulseStops.sort((a, b) => a.label.localeCompare(b.label));

  const meta = {
    source: 'GRTC static GTFS',
    publisher: 'Greater Richmond Transit Company',
    publisherUrl: 'https://www.ridegrtc.com',
    gtfsZipUrl: GTFS_URL,
    feedVersion,
    feedEndDate: feedEnd,
    generatedAt: new Date().toISOString(),
    stopCountRichmondBox: inBox.length,
    pulseStopCount: pulseStops.length,
    bounds: BOUNDS,
    licenseNote:
      'Use subject to GRTC GTFS license. Data is scheduled service only — not live vehicle tracking.',
  };

  writeFileSync(join(DATA, 'grtcGtfsMeta.json'), JSON.stringify(meta, null, 2));
  writeFileSync(join(DATA, 'grtcStopsRichmond.json'), JSON.stringify(inBox, null, 2));
  writeFileSync(join(DATA, 'grtcPulseStops.json'), JSON.stringify(pulseStops, null, 2));

  console.log('Wrote grtcGtfsMeta.json, grtcStopsRichmond.json (%d stops), grtcPulseStops.json (%d)', inBox.length, pulseStops.length);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: whats-on-rva/scripts/import_grtc_gtfs.py
"""
Download GRTC static GTFS and write trimmed JSON (stdlib only).
Run: python scripts/import_grtc_gtfs.py

Also available: npm run import-grtc (Node + adm-zip) — same outputs.
"""
from __future__ import annotations

import csv
import io
import json
import urllib.request
import zipfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "src" / "data"
GTFS_URL = "https://www.ridegrtc.com/wp-content/uploads/2025/02/gtfs.zip"

BOUNDS = {"south": 37.465, "north": 37.615, "west": -77.585, "east": -77.375}


def in_bounds(lat: float, lng: float) -> bool:
    return BOUNDS["south"] <= lat <= BOUNDS["north"] and BOUNDS["west"] <= lng <= BOUNDS["east"]


def main() -> None:
    DATA.mkdir(parents=True, exist_ok=True)
    print("Fetching", GTFS_URL)
    raw = urllib.request.urlopen(GTFS_URL, timeout=120).read()
    zf = zipfile.ZipFile(io.BytesIO(raw))

    def read_txt(name: str) -> str:
        return zf.read(name).decode("utf-8", errors="replace")

    stops_rows = list(csv.DictReader(io.StringIO(read_txt("stops.txt"))))
    stops = []
    for r in stops_rows:
        try:
            lat = float(r["stop_lat"])
            lng = float(r["stop_lon"])
        except (KeyError, ValueError):
            continue
        stops.append(
            {
                "id": str(r.get("stop_id", "")),
                "code": str(r.get("stop_code", "") or ""),
                "name": (r.get("stop_name") or "Stop").strip(),
                "lat": lat,
                "lng": lng,
            }
        )

    in_box = [s for s in stops if in_bounds(s["lat"], s["lng"])]

    routes = list(csv.DictReader(io.StringIO(read_txt("routes.txt"))))
    pulse_route_ids: set[str] = set()
    for r in routes:
        rid = r.get("route_id", "")
        short = (r.get("route_short_name") or "").strip()
        long = (r.get("route_long_name") or "").upper()
        if rid == "BRT" or short == "BRT" or "PULSE" in long:
            pulse_route_ids.add(rid)

    trips = list(csv.DictReader(io.StringIO(read_txt("trips.txt"))))
    brt_trip_ids = {t["trip_id"] for t in trips if t.get("route_id") in pulse_route_ids}
    print("BRT/Pulse trips:", len(brt_trip_ids))

    pulse_stop_ids: set[str] = set()
    with zf.open("stop_times.txt") as f:
        text = io.TextIOWrapper(f, encoding="utf-8")
        reader = csv.DictReader(text)
        for row in reader:
            if row.get("trip_id") in brt_trip_ids:
                sid = row.get("stop_id")
                if sid:
                    pulse_stop_ids.add(str(sid))

    stop_by_id = {s["id"]: s for s in stops}
    pulse_stops = []
    for sid in sorted(pulse_stop_ids, key=lambda x: int(x) if x.isdigit() else x):
        s = stop_by_id.get(sid)
        if s and in_bounds(s["lat"], s["lng"]):
            pulse_stops.append(
                {
                    "id": f"pulse-{s['id']}",
                    "stopId": s["id"],
                    "code": s["code"],
                    "label": f"Pulse (BRT) · {s['name']}",
                    "lat": s["lat"],
                    "lng": s["lng"],
                }
            )

    feed_rows = list(csv.DictReader(io.StringIO(read_txt("feed_info.txt"))))
    fi = feed_rows[0] if feed_rows else {}

    meta = {
        "source": "GRTC static GTFS",
        "publisher": "Greater Richmond Transit Company",
        "publisherUrl": "https://www.ridegrtc.com",
        "gtfsZipUrl": GTFS_URL,
        "feedVersion": fi.get("feed_version", "unknown"),
        "feedEndDate": fi.get("feed_end_date", ""),
        "generatedAt": __import__("datetime").datetime.now(__import__("datetime").timezone.utc).isoformat(),
        "stopCountRichmondBox": len(in_box),
        "pulseStopCount": len(pulse_stops),
        "bounds": BOUNDS,
        "licenseNote": "Use subject to GRTC GTFS license. Scheduled service only — not live vehicle tracking.",
    }

    (DATA / "grtcGtfsMeta.json").write_text(json.dumps(meta, indent=2), encoding="utf-8")
    (DATA / "grtcStopsRichmond.json").write_text(json.dumps(in_box, indent=2), encoding="utf-8")
    (DATA / "grtcPulseStops.json").write_text(json.dumps(pulse_stops, indent=2), encoding="utf-8")
    print(
        f"Wrote grtcGtfsMeta.json, grtcStopsRichmond.json ({len(in_box)}), grtcPulseStops.json ({len(pulse_stops)})"
    )


if __name__ == "__main__":
    main()
